src/main/java/org/inferred/freebuilder/processor/property/BiMapProperty.java

```diff
@@ -0,0 +1,406 @@
+      code.add(" * @throws NullPointerException if ");
+      if (unboxedKeyType.isPresent()) {
+        code.add("{@code value} is");
+      } else if (unboxedValueType.isPresent()) {
+        code.add("{@code key} is");
+      } else {
+        code.add("either {@code key} or {@code value} are");
+      }
+      code.add(" null\n");
+    }
+    code.addLine(" */");
+    addPutAnnotations(code);
+    code.addLine(
+        "public %s %s(%s key, %s value) {",
+        datatype.getBuilder(),
+        putMethod(property),
+        unboxedKeyType.orElse(keyType),
+        unboxedValueType.orElse(valueType)
+    );
+    if (!unboxedKeyType.isPresent()) {
+      code.addLine("  %s.requireNonNull(key);", Objects.class);
+    }
+    if (!unboxedValueType.isPresent()) {
+      code.addLine("  %s.requireNonNull(value);", Objects.class);
+    }
+    code.addLine("  %s.put(key, value);", property.getField())
+        .addLine("  return (%s) this;", datatype.getBuilder())
+        .addLine("}");
+  }
+
+  private void addPutAll(SourceBuilder code) {
+    code.addLine("")
+        .addLine("/**")
+        .addLine(" * Copies all of the mappings from {@code bimap} to the bimap to be returned from")
+        .addLine(" * %s.", datatype.getType().javadocNoArgMethodLink(property.getGetterName()))
+        .addLine(" *")
+        .addLine(" * @return this {@code %s} object", datatype.getBuilder().getSimpleName())
+        .addLine(" * @throws NullPointerException if {@code bimap} is null or contains a")
+        .addLine(" *     null key or value")
+        .addLine(" */");
+    addAccessorAnnotations(code);
+    code.addLine(
+        "public %s %s(%s<? extends %s, ? extends %s> bimap) {",
+        datatype.getBuilder(),
+        putAllMethod(property),
+        BiMap.class,
+        keyType,
+        valueType
+    )
+        .addLine("  for (%s<? extends %s, ? extends %s> entry : bimap.entrySet()) {",
+            Map.Entry.class, keyType, valueType
+        )
+        .addLine("    %s(entry.getKey(), entry.getValue());", putMethod(property))
+        .addLine("  }")
+        .addLine("  return (%s) this;", datatype.getBuilder())
+        .addLine("}");
+  }
+
+  private void addRemoveKey(SourceBuilder code) {
+    code.addLine("")
+        .addLine("/**")
+        .addLine(" * Removes the mapping for {@code key} from the bimap to be returned from")
+        .addLine(
+            " * %s, if one is present.",
+            datatype.getType().javadocNoArgMethodLink(property.getGetterName())
+        )
+        .addLine(" *")
+        .addLine(" * @return this {@code %s} object", datatype.getBuilder().getSimpleName());
+    if (!unboxedKeyType.isPresent()) {
+      code.addLine(" * @throws NullPointerException if {@code key} is null");
+    }
+    code.addLine(" */")
+        .addLine(
+            "public %s %s(%s key) {",
+            datatype.getBuilder(),
+            "removeKey" + property.getCapitalizedName(),
```

Comment:
`"removeKeyFrom" + ...` seems more readable — also, should live in BuilderMethods.

Reply:
The "from" reads much better, but doesn't the method currently live in BuilderMethods?

Reply:
You'll need to add a new static method, `removeKeyFromMethod`, to `BuilderMethods.java`, so that you're not writing out the formula `"removeKeyFrom" + x` everywhere.